src/main/java/net/neoforged/neoforge/transfer/item/ItemResource.java

```diff
@@ -0,0 +1,126 @@
+package net.neoforged.neoforge.transfer.item;
+
+import com.mojang.serialization.Codec;
+import java.util.Optional;
+import net.minecraft.core.Holder;
+import net.minecraft.core.component.DataComponentHolder;
+import net.minecraft.core.component.DataComponentMap;
+import net.minecraft.core.component.DataComponentPatch;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.network.RegistryFriendlyByteBuf;
+import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.level.ItemLike;
+import net.neoforged.neoforge.transfer.Resource;
+
+/**
+ * Immutable combination of an {@link Item} and data components.
+ * Similar to an {@link ItemStack}, but immutable and without amount information.
+ */
+// TODO: probably a "hot take" but I actually prefer ItemVariant :P
+public final class ItemResource implements Resource, DataComponentHolder {
```

Comment:
It would be useful to have the following methods (potentially even having some of them be on the `Resource` level instead of just for `ItemResource`):
- `getMaxSize` (to get the max size of the resource when in stack form, would allow removing a good number of the TODOs about allocation)
- `isComponentsPatchEmpty` (mimics the `ItemStack` method just so that fewer checks have to be done)
- A helper method to get an "item only" variant. For example: `isComponentsPatchEmpty() ? this : of(getItemHolder())`

Reply:
First two definitely. What's the use case for the last one?

Reply:
I'm sure there's some cases where a `plainCopy()` would be helpful. It's just sugar over `ItemResource.of(res.getItem())`, but a bit less lossy when you want to discard the components.